Add ERC: Multistep Contract Ownership (#1212)
ethereum/ERCs: ERCS/erc-8023.md

---
eip: 8023
title: Multi-step Contract Ownership
description: Multi-step contract ownership for secure smart contract ownership management
author: David Kim (@PowerStream3604)
discussions-to: https://ethereum-magicians.org/t/erc-8023-multi-step-contract-ownership/25475
status: Draft
type: Standards Track
category: ERC
created: 2025-09-16
---

## Abstract

> We define a Multi-Stepped contract ownership interface for a more secure contract ownership management. This makes the ownership transfer into 3 distinct steps. With the first 2 steps, performed by the original owner (initiate → confirm) and the remaining 1 step performed by the new owner (accept).

Reviewer: ```suggestion
We define a multi-stepped contract ownership interface for more secure contract ownership management. This makes the ownership transfer into 3 distinct steps. With the first 2 steps, performed by the original owner (initiate → confirm) and the remaining 1 step performed by the new owner (accept).
```

> We enforce an optional time window between initiate and confirm stage to give additional room for review, and make ownership key compromise scenario less fatal.

Reviewer: ```suggestion
We enforce an optional time window between the initiate and confirm stages to give additional room for review, and make ownership key compromise scenarios less fatal.
```

## Motivation

Ownership management is crucial in on-chain security and a significant portion of the security assumptions of defi protocols, smart contract wallets and on chain utilities rely on the contract ownership.

> The single-step `transferOwnership()` style ownership mechanism has been in the industry for a long time, by [ERC-173](./eip-173.md) and has been used ubiquitously as an industry standard. As the industry evolve and attacks get more sophisticated there is a strong need for a multi-step, time gated ownership management to enhance the ecosystem’s contract ownership to be more reviewable, stoppable, thorough and secure.

Reviewer: ```suggestion
The single-step `transferOwnership()` style ownership mechanism has been in the industry for a long time (eg. [ERC-173](./eip-173.md)), and has been used ubiquitously as an industry standard. As the industry evolves and attacks get more sophisticated there is a strong need for a multi-step, time gated ownership management process to enhance the ecosystem’s contract ownership to be more reviewable, stoppable, thorough and secure.
```

## Specification

The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

> 

Reviewer: Might want to mention something about implementing `MultiStepOwnable`:

```suggestion
A multi-step ownable contract MUST implement the following interface:
```

	
	/// @dev returns the pending owner of the account.
	/// pending owner should not have any authority/access/right.
	function pendingOwner() external view returns (address);
	
	/// @dev returns the pre-confirmed owner of the account.
	/// pre-confirmed owner should not have any authority/access/right.
	function preConfirmedOwner() external view returns (address);
	
	/// @dev returns the ownership transfer buffer time (in seconds).
	/// the buffer is enforced between initiation <> confirmation of ownership transfer. 
	/// the standard does not enforce the value range. it is highly recommended to be between 2 <> 14 days.
	function getOwnershipTransferBuffer() external view returns (uint256);
}
```

The `MultiStepOwnable` contract MAY implement the `UpdateableOwnershipTransferBuffer` interface to enable buffer period modification.

The contract MUST update the buffer period with initiation and then confirmation after the existing buffer period.
If this behavior is not enforced, the security of `ownershipTransferBuffer` could break during owner key compromise scenario.

```solidity
/// @title UpdateableOwnershipTransferBuffer. Extension of MultiStepOwnable.

> contract UpdateableOwnershipTransferBuffer {

Reviewer: Interface?

	/// @dev initiates the update of ownership transfer time buffer.
	function initiateOwnershipBufferUpdate(uint256 newBuffer) external;
	
	/// @dev confirms the update of ownership transfer time buffer.
	/// confirmation SHOULD revert if existing time buffer did not pass since
	/// initiation of ownership transfer time buffer.
	function confirmOwnershipBufferUpdate(uint256 newBuffer) external;
}
```

## Rationale

> We define the main objective of this standard to make ownership more secure and handled in a multi-stepped approach that enables the operation to be conducted with more caution and lesser operational mistakes, while being immune to potential scam attacks.

Reviewer: The "main objective" of a proposal should live in the Motivation section. The Rationale section should be used to explain choices made within the proposal itself. My pet analogy is:

> Motivation: We need to build a shed because...
> Rationale: We chose to paint the shed red because...

**3. Ability to rollback ownership transfer, during the transfer process:** Whether through an operational mistake or private key leak of owner account, or other reasons, the ability to rollback ownership transfer within the time buffer highly increases the security and operational burden.

Even in the extreme case of owner private key leak, if the buffer time is set enough, the original owner can earn time to evacuate the funds from the protocol, and possibly prohibit ownership transfer through DoS of ownership (attack → initiate , original owner(defender) → re initiate. which will reset the time back to 0).

**4. Simplicity:** `MultistepOwnable` is targeted to be a simple contract given the diverse use cases and scenarios it could be applied. The process for ownership transfer is concise but thorough enough to allow owners review each step. This is the rationale behind making the ownership transfer time buffer and buffer time update capability optional.

## Backwards Compatibility

> TBD

Reviewer: ```suggestion
TBD <!-- TODO -->
```
